crates/autopilot/src/decoded_settlement.rs

```diff
@@ -254,6 +254,68 @@ impl DecodedSettlement {
             }
         })
     }
+
+    fn fee(
+        &self,
+        external_prices: &ExternalPrices,
+        order: &OrderExecution,
+        trade: &DecodedTrade,
+    ) -> Option<U256> {
+        let solver_fee = match &order.executed_solver_fee {
+            Some(solver_fee) => u256_to_big_rational(solver_fee),
+            None => {
+                // this should be a partial limit order
+                if !trade.flags.partially_fillable() {
+                    tracing::warn!("missing solver fee for non partial fee order");
+                    return None;
+                }
+
+                // get uniform prices
+                let sell_index = self
+                    .tokens
+                    .iter()
+                    .position(|token| token == &order.sell_token)?;
+                let buy_index = self
+                    .tokens
+                    .iter()
+                    .position(|token| token == &order.buy_token)?;
+                let uniform_sell_price = self.clearing_prices.get(sell_index).cloned()?;
+                let uniform_buy_price = self.clearing_prices.get(buy_index).cloned()?;
+
+                // get executed(adjusted) prices
+                let sell_index = trade.sell_token_index.as_u64() as usize;
+                let buy_index = trade.buy_token_index.as_u64() as usize;
+                let adjusted_sell_price = self.clearing_prices.get(sell_index)?;
+                let adjusted_buy_price = self.clearing_prices.get(buy_index)?;
+
+                // the logic is opposite to the code in function `custom_price_for_limit_order`
+                let fee = match trade.flags.order_kind() {
+                    OrderKind::Buy => {
+                        let executed_sell_amount = trade
+                            .executed_amount
+                            .checked_mul(uniform_buy_price)?
+                            .checked_div(uniform_sell_price)?;
+                        adjusted_buy_price.checked_sub(executed_sell_amount)?
```

Comment:
Same concern as with the `sell` version.
This would be my suggestion:
```rust
let required_sell_amount = trade.executed_amount * adjusted_buy_price / adjusted_sell_price;
let required_sell_amount_with_ucp = trade.executed_amount * uniform_buy_price / uniform_sell_price;
let fee = required_sell_amount - required_sell_amount_with_ucp;
```

Reply:
Commited. Related to https://github.com/cowprotocol/services/pull/1425#discussion_r1198837899